File: solutions/c/rail-fence-cipher/1/rail_fence_cipher.c
#include "rail_fence_cipher.h"

#include <string.h>
#include <malloc.h>
#include <stdbool.h>

char *encode(char *text, size_t rails){
    const size_t len = strlen(text);
    const size_t max_step = 2*(rails - 1);
    char *ciphertext = (char*)calloc(len+1, sizeof(char));
    size_t dest = 0;
    for(size_t rail = 0; rail<rails; ++rail){
         size_t step = ( 0==rail ) ? max_step : 2*rail;
         for(size_t text_pos = rail; text_pos <len; 
              step = (max_step == step) ? max_step : max_step - step,   
             text_pos += step, ++dest ){             
              ciphertext[dest] = text[text_pos];   
         }             
    }
    return ciphertext;
}

char *decode(char *ciphertext, size_t rails){
    const size_t len = strlen(ciphertext);
    const size_t max_step = 2*(rails - 1);
    char *cleartext = (char*)calloc(len+1, sizeof(char));
    size_t dest = 0;
    for(size_t rail = 0; rail<rails; ++rail){
         size_t step = ( 0==rail ) ? max_step : 2*rail;
         for(size_t text_pos = rail; text_pos <len; 
              step = (max_step == step) ? max_step : max_step - step,   
             text_pos += step, ++dest ){             
               cleartext[text_pos] = ciphertext[dest];   
         }             
    }
    return cleartext;
}

File: solutions/python/rail-fence-cipher/1/rail_fence_cipher.py
def encode(message, rails):
    return coder(message, rails, "encode")


def decode(encoded_message, rails):
    return coder(encoded_message, rails, "decode")

def coder(text, rails, mode):
    max_step = 2*(rails-1)
    output=["\0" for i in range(0, len(text))]
    position = 0
    for the_rail in range(0,rails):
        step = 2*the_rail
        if 0==the_rail:
            step = max_step
        text_pos = the_rail
        while text_pos< len(text):
            if mode=="encode":
                output[position] = text[text_pos] 
            elif mode=="decode":
                output[text_pos] = text[position]
            else: 
                raise ValueError("Invalid mode.")
            if step!=max_step:
                step = max_step - step
            text_pos += step
            position +=1
    return ''.join(output)  